libraries/AP_OSD/AP_OSD_ParamScreen.cpp

```diff
@@ -0,0 +1,449 @@
+void AP_OSD_ParamScreen::update_state_machine()
+{
+    const uint32_t now = AP_HAL::millis();
+    if ((now - _transition_start_ms) < _transition_timeout_ms) {
+        return;
+    }
+
+    const Event ev = map_rc_input_to_event();
+    // only take action on transitions
+    if (ev == Event::NONE && ev == _last_rc_event) {
+        return;
+    }
+
+    debug("update_state_machine(%s)\n", event_names[int(ev)]);
+
+    _transition_start_ms = now;
+    if (ev == _last_rc_event) {
+        _transition_timeout_ms = OSD_HOLD_BUTTON_PRESS_DELAY;
+    } else {
+        _transition_timeout_ms = osd->button_delay_ms;
+    }
+    _last_rc_event = ev;
+
+    // if we were armed then there is no selected parameter - so find one
+    if (_selected_param == 0) {
+        _selected_param = 1;
+        for (uint8_t i = 0; i < NUM_PARAMS && !params[_selected_param-1].enabled; i++) {
+            _selected_param++;
+        }
+    }
+
+    switch (ev) {
+    case Event::MENU_ENTER:
+        _modify_param = true;
+        break;
+    case Event::MENU_UP:
+        if (_modify_param) {
+            modify_parameter(_selected_param, ev);
+        } else {
+            _selected_param = (_selected_param + NUM_PARAMS - 2) % NUM_PARAMS + 1;
+            // skip over parameters that are not enabled
+            for (uint8_t i = 0; i < NUM_PARAMS && !params[_selected_param-1].enabled; i++) {
+                _selected_param = (_selected_param + NUM_PARAMS - 2) % NUM_PARAMS + 1;
+            }
+        }
+        break;
+    case Event::MENU_DOWN:
+        if (_modify_param) {
+            modify_parameter(_selected_param, ev);
+        } else {
+            _selected_param = _selected_param % NUM_PARAMS + 1;
+            // skip over parameters that are not enabled
+            for (uint8_t i = 0; i < NUM_PARAMS && !params[_selected_param-1].enabled; i++) {
+                _selected_param = _selected_param % NUM_PARAMS + 1;
+            }
+        }
+        break;
+    case Event::MENU_EXIT:
+        if (_modify_param) {
+            save_parameter(_selected_param);
+            _modify_param = false;
+        }
+        break;
+    case Event::NONE:
+        break;
+    }
+}
+
+void AP_OSD_ParamScreen::draw(void)
+{
+    if (!enabled || !backend) {
+        return;
+    }
+
+    // first update the state machine
+    if (!AP::arming().is_armed()) {
+        update_state_machine();
+    } else {
+        _selected_param = 0;
+    }
+
+    for (uint8_t i = 0; i < NUM_PARAMS; i++) {
+        AP_OSD_ParamSetting n = params[i];
+        if (n.enabled) {
+            draw_parameter(n._number, n.xpos, n.ypos);
+        }
+    }
+}
+
+// pre_arm_check - returns true if all pre-takeoff checks have completed successfully
+bool AP_OSD::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) const
+{
+    // currently in the OSD menu, do not allow arming
+    if (!is_arming_allowed()) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "In OSD menu");
```

Comment:
This is making an assumption about why `is_arming_allowed()` said no.

Reply:
Fixed the name